### PR TITLE
Additional logging for the file provisioner and a fix for the master branch of the element interpolation bug

### PR DESCRIPTION
**Pull request by knuckolls**
Will link to this from the issue that explains it.

### CODE REVIEW

File: builtin/provisioners/file/resource_provisioner.go [hunk 1]
```diff
@@ -22,6 +22,9 @@ func (p *ResourceProvisioner) Apply(
 		return err
 	}
 
+	log.Printf("private_ip attribute: %#v", s.Attributes["private_ip"])
+	log.Printf("connection:host attribute: %#v", s.Ephemeral.ConnInfo["host"])
```
Comment (mitchellh) [inline builtin/provisioners/file/resource_provisioner.go]: Can you prefix these with `[DEBUG]`? We haven't enabled filtering yet but we plan to.

**Comment by mitchellh** [comment]
For the wrong element bug, we're going to really need a test to make sure we don't regress it when we merge #785. Can you add that?

**Comment by knuckolls** [comment]
Sure I'll look into adding the test. I also went ahead and pushed the debug prefix.

**Comment by knuckolls** [comment]
Thinking through the implementation of this test and reading through the existing ones. Is there a particular reason you guys chose to wrap if the index is out of bounds?

https://github.com/hashicorp/terraform/blob/master/config/interpolate_funcs_test.go#L152

**Comment by brycekahle** [comment]
> Is there a particular reason you guys chose to wrap if the index is out of bounds?

@knuckolls I'm speculating, but I'm guessing it is for the case when the list you are applying the `element` function to has fewer elements than the count of the resource. 

A good example is subnets in AWS. If you are launching 9 instances, but only run them in 3 subnets (corresponding to AZs), the auto-modulus is nice (e.g. `${element(aws_subnet.blah.*.id, count.index)}`. This could be solved another way by having operators and a list length() function in HCL. That is a much larger task and I'm not sure it is something hashicorp even wants.

**Comment by phinze** [comment]
Marking as `waiting-for-response` since we're waiting on tests here. :+1:

**Comment by mitchellh** [comment]
@knuckolls Sorry to do this to you, but could you resubmit this PR rebased against master. Also, could you split this PR into two? I'd like to get the additional logging in but want to wait on the other core change. Thanks!

**Pull request closed without merging**